File: backend/app/schemas/local_sensor.py
# PacketArch — OT Traffic Simulation Platform
# Copyright (c) 2026 Rocky Smith <rocky.d.smith@proton.me>
# Licensed under GPL-3.0. See LICENSE at the repo root.
"""Local sensor lab schemas.

Mirrors the CML build-lab schema style (schemas/cml.py) but for an on-box,
app-managed (agent + CV sensor + virtual SPAN) lab provisioned by the
privileged host-agent.
"""

from pydantic import BaseModel, Field


# --- host capability status -------------------------------------------------

class LocalHostStatusResponse(BaseModel):
    """Whether the host-agent capability is present and reachable."""

    available: bool = Field(..., description="Shared state volume mounted (host-agent wired)")
    host_agent_seen: bool = Field(..., description="Host-agent has initialized its state dirs")
    message: str


# --- build ------------------------------------------------------------------

class LocalLabBuildRequest(BaseModel):
    """Schema for building a local agent + CV-sensor lab on the host.

    The CV sensor is auto-provisioned via the Cyber Vision API (Settings >
    Cyber Vision must be configured) — no compose paste required.
    """

    name: str = Field(..., min_length=1, max_length=128, description="Name for the new local lab")
    agent_name: str | None = Field(
        default=None,
        max_length=255,
        description="Name for the PacketArch agent; default derived from the lab slug",
    )


class LocalLabBuildResponse(BaseModel):
    """Result of a build request (token shown once, like CML)."""

    success: bool
    message: str
    lab_id: str | None = None
    slug: str | None = None
    agent_id: str | None = None
    agent_token: str | None = Field(default=None, description="Agent token (shown only once)")
    sensor_serial: str | None = None
    state: str = "pending"
    warnings: list[str] = Field(default_factory=list)


# --- listing / detail -------------------------------------------------------

class LocalLabItem(BaseModel):
    """A local sensor lab with its live status."""

    lab_id: str
    name: str
    slug: str
    state: str
    status_detail: str | None = None
    agent_id: str | None = None
    agent_name: str | None = None
    agent_status: str | None = None
    sensor_serial: str | None = None
    gen_if: str
    mon_if: str
    # Live fields merged from the host-agent status file (best-effort).
    stage: str | None = None
    percent: int | None = None
    resources: dict | None = None


class LocalLabListResponse(BaseModel):
    items: list[LocalLabItem]


class LocalLabTeardownResponse(BaseModel):
    success: bool
    message: str

File: frontend/src/api/localSensor.ts
/*
 * PacketArch — OT Traffic Simulation Platform
 * Copyright (c) 2026 Rocky Smith <rocky.d.smith@proton.me>
 * Licensed under GPL-3.0. See LICENSE at the repo root.
 */
/**
 * Local sensor lab API — app-managed (agent + CV sensor + virtual SPAN) labs
 * on the PacketArch host itself. Mirrors api/cml.ts.
 */

import apiClient from './client';

export interface LocalHostStatus {
  available: boolean;
  host_agent_seen: boolean;
  message: string;
}

export interface LocalLabBuildRequest {
  name: string;
  agent_name?: string | null;
}

export interface LocalLabBuildResponse {
  success: boolean;
  message: string;
  lab_id: string | null;
  slug: string | null;
  agent_id: string | null;
  agent_token: string | null;
  sensor_serial: string | null;
  state: string;
  warnings: string[];
}

export interface LocalLabResources {
  veth?: boolean;
  sensor_running?: boolean;
  agent_running?: boolean;
}

export interface LocalLabItem {
  lab_id: string;
  name: string;
  slug: string;
  state: string;
  status_detail: string | null;
  agent_id: string | null;
  agent_name: string | null;
  agent_status: string | null;
  sensor_serial: string | null;
  gen_if: string;
  mon_if: string;
  stage: string | null;
  percent: number | null;
  resources: LocalLabResources | null;
}

export const localSensorApi = {
  getHostStatus: async (): Promise<LocalHostStatus> => {
    const response = await apiClient.get<LocalHostStatus>('/api/v1/local-sensor/host-status');
    return response.data;
  },

  getLabs: async (): Promise<{ items: LocalLabItem[] }> => {
    const response = await apiClient.get<{ items: LocalLabItem[] }>('/api/v1/local-sensor/labs');
    return response.data;
  },

  getLab: async (labId: string): Promise<LocalLabItem> => {
    const response = await apiClient.get<LocalLabItem>(`/api/v1/local-sensor/${labId}`);
    return response.data;
  },

  build: async (request: LocalLabBuildRequest): Promise<LocalLabBuildResponse> => {
    const response = await apiClient.post<LocalLabBuildResponse>('/api/v1/local-sensor/build', request);
    return response.data;
  },

  teardown: async (labId: string): Promise<{ success: boolean; message: string }> => {
    const response = await apiClient.post<{ success: boolean; message: string }>(
      `/api/v1/local-sensor/${labId}/teardown`,
    );
    return response.data;
  },
};

export default localSensorApi;
